File: Julia/3 - QuickSort.jl
function qsort(a)
    b = qsort_kernel(a,1,length(a));
    return b;
end

function qsort_kernel(a,lo,hi)
    i = lo;
    j = hi;
    while i<=j
        pivor = a[trunc(Int,(lo+hi)/2)];
        while i<=j
            while a[i] < pivor && i<hi
                i = i + 1;
            end
            while (a[j] > pivor && j>lo)
                j = j - 1;
            end
            if i<=j
                t = a[i];
                a[i] = a[j];
                a[j] = t;
                i = i + 1;
                j = j - 1;
            end
        end
        if j>lo
            a = qsort_kernel(a,lo,j);
        end
        if i<hi
            a = qsort_kernel(a,i,hi);
        end
    end
    return a;
end

function mainn()
    n = 1000;
    v = rand(n);
    resp = qsort(v);

    for i=1:n
            println(resp[i])
    end
end

println("TEMPO: ",@elapsed mainn())

File: C/3 - Quick_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
int n = 1000;

void quicksort (int a[]) {
    qsort_kernel(&a,1,n+1);
}

void qsort_kernel(int a[], int lo, int hi) {
    int i = lo, j = hi, t;
    int pivor = a[(lo+hi)/2];
    while(i<=j) {
        while (i<=j) {
            while (a[i]<pivor && i<hi) {
                i++;
            }
            while (a[j]>pivor && j>lo) {
                j--;
            }
            if (i<=j) {
                t = a[i];
                a[i] = a[j];
                a[j] = t;
                i++;
                j--;
            }
        }
        if (j>lo) {
            qsort_kernel(a,lo,j);
        }
        if (i<hi) {
            qsort_kernel(a,i,hi);
        }
        return;
    }
}

int main() {
    int i;
    int a[n];
    for(i=0;i<n;i++){
        a[i] = rand()%n;
    }
    for(i=0;i<n;i++){
        printf("%d ",a[i]);
    }
    printf("\n\n");
    quicksort(&a);
    for(i=0;i<n;i++){
        printf("%d ",a[i]);
    }
    printf("\n\n");
    return 0;
}
